### PR TITLE
update webpack to 4.2.0

### PR DESCRIPTION
**Pull request by CaioMatias**
- minor version pump into dependencies
- small change in client entry so it can accept other functions later (preloadCSS, initOffline)
- config fixes for webpack version pump

### CODE REVIEW

File: config/webpack.dev.js [hunk 1]
```diff
@@ -1,23 +1,33 @@
...
+            'process.env.NODE_ENV': JSON.stringify(
+                process.env.NODE_ENV || 'development'
+            ),
+        }),
         new webpack.HotModuleReplacementPlugin(),
     ],
     module: {
         rules: [],
     },
+    node: {
```
Comment (unruffledBeaver) [inline config/webpack.dev.js]: Why do we need to create these empty objects? Are there things running in browser that expect node globals?

File: config/webpack.dev.js [hunk 2]
```diff
@@ -1,23 +1,33 @@
...
+        dgram: 'empty',
+        fs: 'empty',
+        net: 'empty',
+        tls: 'empty',
+        child_process: 'empty',
+    },
+    performance: {
+        hints: false,
```
Comment (unruffledBeaver) [inline config/webpack.dev.js]: I havent run into one yet but how loud are these warnings? Seems like something we want enabled in dev so webpack will notify us of things like large bundles

**Comment by rynocouse** [comment]
I think we should move (inline) all build configs that webpack uses into webpack (.babelrc, postcss.conf.js, etc)

Few things I'd like us to consider:
 - Remove the magical `merge.smart` (one more set of docs I have to read to understand what's happening w/ the webpack config)
 - Building a single webpack file and overriding things as needed for dev based on NODE_ENV. (this could be a seperate PR)

File: config/webpack.base.js [hunk 3]
```diff
@@ -17,6 +16,18 @@ module.exports = {
             'redux-thunk',
         ],
     },
+    optimization: {
```
Comment (rynocouse) [inline config/webpack.base.js]: Do we need to specify our venor code splitting anymore? My understanding is webpack 4 now automagically splits out vendors as needed by default.

**Comment by unruffledBeaver** [comment]
I am actually partial to splitting configs out by what they handle i.e. JS, CSS, Assets, etc and merging them in the configs right before exporting so they are more approachable and its easy to see how a given import type is processed rather than a monolithic config for all import types.

**Comment by rynocouse** [comment]
I'll take a "monolithic" config over the current magic. Looking at webpack.base there's no way for me to know it's using postcss.conf.js or .babelrc. It's by each individual loaders convention. You have to sift through each loaders documentation to figure out they're loading a arbitrarily named config file in the project root. I'm cool keeping them seperate if we explicitly import them and put them in a folder. Also another issue we need to address is we are defining our browserlist in two places

**Comment by unruffledBeaver** [comment]
100% agree on moving all configs that webpack uses into webpack and doing away with `rc` files. My approach is a little more similar to the old gulp setup in the you would have separate files for each loader type and that file would have both dev and prod configs being setup and exported and then in an `index.js` we would merge them into one config for each environment.

I agree right now its hard to follow whats happening because its both a monolith and we merge those monoliths to create environment configs.

For all intents and purposes if this PR works with our code as is, I say we merge once comments have been addressed and take a bigger refactor effort into another branch

**Comment by rynocouse** [comment]
I agree, If this works let's merge.

**Comment by stevescavo** [comment]
@rynocouse @unruffledBeaver: we'll need to do some decent amount of testing on this upgrade. If it check out, I'm okay with merging it in as a bump to 4.x is nice. We can riff on our overall webpack config in a separate PR.

File: config/webpack.base.js (hunk 3 above)
Comment (CaioMatias) [reply config/webpack.base.js]: I tried without an explicit vendor chunk definition and it put all together in the client bundle. Maybe it has a single configuration param to do that and I'm not aware.

File: config/webpack.dev.js (hunk 1 above)
Comment (CaioMatias) [reply config/webpack.dev.js]: I used to have it because some libraries import Node modules but don't use them in the browser. Maybe we don't need it in this boiler, I'll remove it.

File: config/webpack.dev.js (hunk 2 above)
Comment (CaioMatias) [reply config/webpack.dev.js]: It warms mostly on file sizes everytime the bundle updates, it's annoying in development. I think it's something we must have while bundling for production but not everytime you save a file.

Comment (unruffledBeaver) [reply config/webpack.dev.js]: IMO I think this is an important warning as the biggest benefit of Webpack is code splitting. These should be annoying to let us know while deving we should consider splitting a bundle

**Comment by stevescavo** [comment]
I'm unintentionally closed this PR. Reopening.
